weasyl/oauth2.py

```diff
@@ -92,3 +118,72 @@ def get_userid_from_authorization(scopes=['wholesite']):
 __all__ = [
     'get_consumers_for_user', 'revoke_consumers_for_user',
 ]
+
+
+def get_allowed_scopes(user_id):
+    """
+    Get a list of oauth scopes this user is allowed to request
+    :param user_id: the userid of the application owner
+    :return: a list of scopes
+    """
+    allowed = _SCOPES
+    # only trusted individuals should be allowed to use the "wholesite" oauth grant
+    if user_id not in staff.ADMINS | staff.MODS | staff.DEVELOPERS:
+        allowed = [scope for scope in allowed if scope['name'] != 'wholesite']
+    return allowed
+
+
+def register_client(user_id, name, scopes, redirects, homepage):
+    """
+    Register an application as an OAuth2 consumer
+    :param user_id: the user registering this application
+    :param name: the name of the application
+    :param scopes: a list of the scopes registered for this application
+    :param redirects: allowed redirect URIs for this application
+    """
+    if not name.strip():
+        raise WeasylError("nameMissing")
+
+    session = d.connect()
+    new_consumer = orm.OAuthConsumer(
+        clientid=security.generate_key(32),
+        description=name,
+        ownerid=user_id,
+        grant_type="authorization_code",
+        response_type="code",
+        scopes=scopes,
+        redirect_uris=redirects,
+        client_secret=security.generate_key(64),
+        homepage=homepage,
+    )
+    # this doesnt seem right
+    session.begin()
+    session.add(new_consumer)
+    session.commit()
+
+
+def get_registered_applications(user_id):
+    """
+    Return a list of all OAuth2 consumers registered to this account
+    :param user_id:
+    :return:
+    """
+    q = (orm.OAuthConsumer.query
+         .filter_by(ownerid=user_id))
+    return q.all()
+
+
+def remove_clients(user_id, clients):
+    q = (orm.OAuthConsumer.query
+         .filter_by(ownerid=user_id)
+         .filter(orm.OAuthConsumer.clientid.in_(clients)))
+    q.delete(synchronize_session=False)
+
+
+def renew_client_secrets(user_id, clients):
+    for cid in clients:
+        q = (orm.OAuthConsumer.query
+             .filter_by(ownerid=user_id)
+             .filter_by(clientid=cid))
+        q.update({orm.OAuthConsumer.client_secret: security.generate_key(64)},
```

Comment:
Just sanity checking: This is one row per run through the loop? And we do it that way so that SQLAlchemy won't use the same security.generate_key(64) result for each?